torch/_export/serialize.py

```diff
@@ -46,33 +70,440 @@ def _reverse_map(d):
+        return Operator(name=target, version=version)
+    elif target in _SYM_INT_OPS:
+        return Operator(name=f"{target.__module__}.{target.__name__}", version=version)
+    elif isinstance(target, torch._ops.HigherOrderOperator):
+        return Operator(name=target.__name__, version=version)
+    else:
+        return Operator(name=str(target), version=version)
+
+
+def export_call_spec(call_spec: Optional[ep.CallSpec]) -> CallSpec:
+    if call_spec is None:
+        return CallSpec(in_spec="", out_spec="")
+    # TODO(angelayi): spec
+    return CallSpec(in_spec="", out_spec="")
+
+
+def export_signature(sig: Optional[ep.GraphSignature]) -> GraphSignature:
+    if sig is None:
+        return GraphSignature(
+            inputs_to_parameters={},
+            inputs_to_buffers={},
+            user_inputs=[],
+            user_outputs=[],
+            buffers_to_mutate={},
+            backward_signature=None,
+        )
+
+    if bw_sig := sig.backward_signature:
+        backward_signature = BackwardSignature(
+            gradients_to_parameters=bw_sig.gradients_to_parameters,
+            gradients_to_user_inputs=bw_sig.gradients_to_user_inputs,
+            loss_output=bw_sig.loss_output,
+        )
+    else:
+        backward_signature = None
+
+    graph_signature = GraphSignature(
+        inputs_to_parameters=sig.inputs_to_parameters,
+        inputs_to_buffers=sig.inputs_to_buffers,
+        user_inputs=sig.user_inputs,
+        user_outputs=sig.user_outputs,
+        buffers_to_mutate=sig.buffers_to_mutate,
+        backward_signature=backward_signature,
     )
+    return graph_signature
+
+
+def export_state_dict(state_dict: Optional[Dict[str, Any]]) -> bytes:
+    if state_dict is None:
+        return bytes("", encoding="utf8")
+
+    buffer = io.BytesIO()
+    state_dict = dict(state_dict)
+    for name in state_dict:
+        # This is a workaround for backend's tensor deserialization problem:
+        # unpickleTensor() always create a tensor on the device where it was originally saved
+        # This behavior is bad for multi-gpu training, as we wish to directly load the tensor
+        # on the designated device.
+        # For now, we simply move the tensor to cpu before saving.
+        # TODO: this should be fixed by deserialization instead.
+        state_dict[name] = state_dict[name].cpu()
+    torch.save(state_dict, buffer)
+    return buffer.getvalue()
+
+
+def _is_single_tensor_return(target: torch._ops.OpOverload) -> bool:
+    returns = target._schema.returns
+    return len(returns) == 1 and isinstance(returns[0].real_type, torch.TensorType)
+
+
+class Serializer:
+    def __init__(self, op_version: int = 0):
+        self.inputs: List[Argument] = []
+        self.outputs: List[Argument] = []
+        self.nodes: List[Node] = []
+        self.tensor_values: Dict[str, TensorValue] = {}
+        self.sym_int_values: Dict[str, SymInt] = {}
+        self.op_version: int = op_version
```

Comment:
oh interesting, so this is actually an operator set version but not per-operator version. Then this should be the constant along the lifecycle of `Serializer` then. I would suggest move this out from `Serializer` for now. In the end it should come from the version table.

Reply:
agree